media_driver/cmake/linux/media_gen_flags_linux.cmake

```diff
@@ -18,6 +18,9 @@
 # ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 # OTHER DEALINGS IN THE SOFTWARE.
 
+option(ONLY_MEDIA_KERNELS "Make per-Gen options disable only media-kernels (not driver code) if driver code is known to be required for a successful build"
```

Comment:
want to clarify these:
1. it is a build option during per-gen options is not clean. related driver code still could not excluded when the per-gen build option is "OFF"
2. from previous conditions "if(ONLY_MEDIA_KERNELS OR GEN12_TGLLP);xxx"; if the option is "ON", it will always build related media driver.  your expectation should be , if ONLY_MEDIA_KERNELS is ON, but GEN12_TGGLP is OFF, then the media kernel files of TGLLP should not be compiled , but the code of TGLLP will still be compiled, right?
from this perspective, the ONLY_MEDIA_KERNELS is looks confusing, it looks not a good name to be a switch.  it is something like "allow source code compiled even disable the platform",  "ALLOW_REDUNDANT_CODE" is also not good enough, I still have no idea with a better name.  or remove the new definitions, but add comments to describe that current status, GENxx=OFF maybe just disable the kernel files.  both methods looks not so graceful. 
which configuration you cared mostly?  if you just care 1 or 2 platform disabling, maybe we could focus on that , resolve it one by one?
3. your changes with github actions does not include any reference of MEDIA_KERNELS, so, why it will benefit your project?

Reply:
>     1. it is a build option during per-gen options is not clean. related driver code still could not excluded when the per-gen build option is "OFF"

It's true that it's not a clean option. My plan/hope is for the `ONLY_MEDIA_KERNELS` option to be removed once each Gen's driver code can be safely disabled.

>     2. from previous conditions "if(ONLY_MEDIA_KERNELS OR GEN12_TGLLP);xxx"; if the option is "ON", it will always build related media driver.  your expectation should be , if ONLY_MEDIA_KERNELS is ON, but GEN12_TGGLP is OFF, then the media kernel files of TGLLP should not be compiled , but the code of TGLLP will still be compiled, right?

That's exactly right.

>        from this perspective, the ONLY_MEDIA_KERNELS is looks confusing, it looks not a good name to be a switch.  it is something like "allow source code compiled even disable the platform",  "ALLOW_REDUNDANT_CODE" is also not good enough, I still have no idea with a better name.  or remove the new definitions, but add comments to describe that current status, GENxx=OFF maybe just disable the kernel files.  both methods looks not so graceful.
>        which configuration you cared mostly?  if you just care 1 or 2 platform disabling, maybe we could focus on that , resolve it one by one?

I agree, it's not graceful and it's somewhat confusing. I'd kind of suggest that this isn't unexpected given that almost none of the `-DGEN` options currently work if set to `OFF` :)

>     3. your changes with github actions does not include any reference of MEDIA_KERNELS, so, why it will benefit your project?

It would benefit anyone that wishes to reduce the disk usage of the project by disabling media kernels (and code, in the case of Gen8).

The intention behind the `ONLY_MEDIA_KERNELS` option (which is `ON` by default) is to provide a baseline of support for disabling the `GEN` options. Currently most of them, when disabled, cause the build to fail. My strategy is to

1. avoid those build failures with the `ONLY_MEDIA_KERNELS` mechanism (while still providing a simple way to reproduce the build failures for purposes of fixing them)
2. allow significant disk usage savings by disabling the media kernels even in cases where the rest of that Gen's driver support code cannot (yet) be disabled

To that end, the GitHub Actions are there to ensure that each of the `-DGEN=` options does not produce a broken build. In the case of Gen 8, this means disabling the media kernels *and* driver code. For other Gens it means only disabling the media kernels -- until those issues are resolved which will enable us to disable the Gen-specific driver code as well.

I could invert the default of `ONLY_MEDIA_KERNELS` from `ON` to `OFF` so that a build configuration that only specified `-DGEN9=OFF` produces a build failure, and I could add `-DONLY_MEDIA_KERNELS=ON` to the GitHub Action for Gen9, but I don't know why that would be better in any way.


Could I get some clarity on what the concerns are? 

* Is the goal of building the project for a single generation in order to reduce the disk usage from 36 MiB to 18 MiB not considered sufficiently valuable?

* Is the strategy of achieving 90% of the goal by allowing disabling media kernels independently of resolving long standing issues not a good one? (Issue #451 was filed in 2018; similar issues #813 and #1065 were filed later).

I'm unsure whether we're in agreement on the goals, the strategy, both, or neither.

Like if it's just confusion about how the option should be named or how it should behave then, great, we can surely find something better. If there's disagreement about the value in reducing binary size, then we probably need to take a step back and have a different discussion. In between those, if there's disagreement on the strategy of allowing disabling media kernels independently of driver code, then we probably need to figure out a new path.

Reply:
thanks @mattst88,  I just noticed you did not add "ONLY_MEDIA_KERNEL" for gen8, because gen8 is clean. yes, we should not change the value of "ONLY_MEDIA_KERNEL" in  configuration Github Actions.
I have no concern with the commit itself , except the naming. better to "ALLOW_CODE_BESIDES_KERNEL" ?  , "ALLOW_REDUNDANT_CODE" or .... , just like what I said, I have no good idea about it. anyway, I agree the value and the purpose. thanks for your contribution

Reply:
Okay, thank you very much!

I've changed the name to `ENABLE_REQUIRED_GEN_CODE` which I think is better than `ONLY_MEDIA_KERNEL`. I've also added a comment above the option documenting that we plan to remove it when all Gen options can disable the driver code.